src/webui/api/torrentscontroller.cpp

```diff
@@ -1038,3 +1038,78 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+
+    for (const auto &tagRaw : asConst(tags)) {
+        const QString tag {tagRaw.trimmed()};
+        applyToTorrents(hashes, [tag](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->hasTag(tag) && !torrent->addTag(tag))
+                throw APIError(APIErrorType::Conflict, tr("Incorrect tag name"));
+        });
+    }
+}
+
+void TorrentsController::removeTagsAction()
+{
+    checkParams({"hashes"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+
+    for (const auto &tagRaw : asConst(tags)) {
+        const QString tag {tagRaw.trimmed()};
+        applyToTorrents(hashes, [tag](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (torrent->hasTag(tag) && !torrent->removeTag(tag))
+                throw APIError(APIErrorType::Conflict, tr("Incorrect tag name"));
```

Comment:
There's something odd about it... If a torrent has a tag, what can prevent you from deleting it? At least incorrect tag name isn't valid reason.

Reply:
For removal purposes valid tag is any, that exists on current torrent.

Reply:
Didn't you understand me?
If torrent has a tag then this tag has correct name. So At "incorrect tag name" isn't valid reason of possible error here.

>For removal purposes, valid tag is any, that exists on current torrent.

So you only allow the removal of a set of tags that exist in each of these torrents? It looks uncomfortable to me.

Reply:
This whole check is redundant and there should be plain `torrent->removeTag(tag);`. Did I understand you correctly?